### PR TITLE
Fix the converts an empty string to nil.

### LINKED ISSUE #118588: Metadata of immutable secrets cannot be changed when the `.data` has a key with empty value
### What happened?

I have created the following secret in my cluster.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: test
immutable: true
data:
  "foo": ""
  "one": dHdvCg==
```

When I add labels to my secret and try to apply it in the cluster I get the following error 
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: test
  labels:
    "foo": "bar"
immutable: true
data:
  "foo": ""
  "one": dHdvCg==

```

```bash
$ k apply -f secret.yaml
The Secret "test" is invalid: data: Forbidden: field is immutable when `immutable` is set
```

This also happens when I use `kubectl edit` or using the `controller-runtime` go client.

Not sure why, but `kubectl label` works.
```bash
$ k label secret test foo=bar
secret/test labeled
```

### What did you expect to happen?

I expect that the secret can be updated with the new labels since the `data` is not actually changed.

### How can we reproduce it (as minimally and precisely as possible)?

Steps explained above.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-15T13:40:17Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"darwin/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.3", GitCommit:"9e644106593f3f4aa98f8a84b23db5fa378900bd", GitTreeState:"clean", BuildDate:"2023-03-30T06:34:50Z", GoVersion:"go1.19.7", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
Kind
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by HirazawaUi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
converter shouldn't transform a "" into a []byte(nil) but into a []byte{}

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/118588

#### Special notes for your reviewer:
For the detailed discussion, please refer to: https://github.com/kubernetes/kubernetes/pull/118605#issuecomment-1628309912
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by HirazawaUi** [comment]
@wojtek-t  @thockin @apelisse Please help review this PR

**Comment by wojtek-t** [comment]
This LGTM, but I will let @apelisse to also take a look.

/approve

/assign @apelisse

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter_test.go [hunk 1]
```diff
@@ -751,6 +752,10 @@ func TestUnrecognized(t *testing.T) {
 			data: "{\"ff\":[\"abc\"],\"fg\":[123],\"fh\":[true,false]}",
 			obj:  &F{},
 		},
+		{
+			data: "{\"fj\":\"\"}",
+			obj:  &F{},
+		},
```
Comment (apelisse) [inline staging/src/k8s.io/apimachinery/pkg/runtime/converter_test.go:758]: I'm not convinced this is good enough? Did that test fail to pass without the change?

Comment (HirazawaUi) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter_test.go:758]: > I'm not convinced this is good enough? Did that test fail to pass without the change?

This test will fails without the change, The error message is as follows:
```
    converter_test.go:691: Object changed, diff:   &runtime_test.F{
          	... // 7 identical fields
          	H: nil,
          	I: nil,
        - 	J: []uint8{},
        + 	J: nil,
          }
```

**Comment by thockin** [comment]
Thanks for the follow-through @HirazawaUi - this has been a fun bug to track.  After this PR, is the `secret` issue resolved?  If so, we should still add a test case for it.  If not, ...uh oh?

I also approve, but @apelisse holds the LGTM.

/approve

**Comment by HirazawaUi** [comment]
> Thanks for the follow-through @HirazawaUi - this has been a fun bug to track. After this PR, is the `secret` issue resolved? If so, we should still add a test case for it. If not, ...uh oh?
> 
> I also approve, but @apelisse holds the LGTM.
> 
> /approve

Yes, the secret issue is solved. If tests still need to be added, should we add them in [sercet validation](https://github.com/kubernetes/kubernetes/blob/186b02e6f723c63cadc2a262b6df1e62f4e0e5fd/pkg/apis/core/validation/validation_test.go#L18756)?

**Comment by apelisse** [comment]
Thanks
/lgtm

**Comment by pacoxu** [comment]
The failure is a flake https://github.com/kubernetes/kubernetes/issues/119244.

/test pull-kubernetes-unit

**Comment by rfranzke** [comment]
Thank you @HirazawaUi! Do you plan to cherry-pick this PR?

**Comment by HirazawaUi** [comment]
> Thank you @HirazawaUi! Do you plan to cherry-pick this PR?

Okay, I'll do this.

**Pull request merged**